File: csharp/02-never-ending-function/src/order_processor.cs
public record LineItem(decimal Price, int Quantity);
public record Customer(string Email);
public record Address(string Zip);
public record Order(List<LineItem> Items, Customer Customer, Address Address);

public class OrderResult
{
    public string? Error;
    public decimal Total;
    public string Carrier = "";
    public string PaymentStatus = "";
    public string EmailTo = "";
    public string EmailSubject = "";
}

public class OrderProcessor
{
    public OrderResult process_order(Order order)
    {
        if (order.Items == null || order.Items.Count == 0)
        {
            return new OrderResult
            {
                Error = "No items"
            };
        }

        if (order.Customer == null || string.IsNullOrEmpty(order.Customer.Email))
        {
            return new OrderResult
            {
                Error = "Invalid customer"
            };
        }

        if (order.Address == null || string.IsNullOrEmpty(order.Address.Zip))
        {
            return new OrderResult
            {
                Error = "Invalid address"
            };
        }

        var inventory = true;
        foreach (var item in order.Items)
        {
            if (item.Quantity > 100)
            {
                inventory = false;
            }
        }

        if (!inventory)
        {
            return new OrderResult
            {
                Error = "Out of stock"
            };
        }

        decimal subtotal = 0;
        var weight = 0;
        foreach (var item in order.Items)
        {
            subtotal += item.Price * item.Quantity;
            weight += item.Quantity;
        }

        decimal shipping = 0;
        if (subtotal < 25)
        {
            shipping = 5.99m;
        }
        else if (subtotal < 50)
        {
            shipping = 3.99m;
        }

        var total = subtotal + subtotal * 0.07m + shipping;
        var carrier = total > 100 ? "UPS" : "USPS";
        var status = total > 1000 ? "pending_review" : "approved";

        return new OrderResult
        {
            Total = Math.Round(total, 2),
            Carrier = carrier,
            PaymentStatus = status,
            EmailTo = order.Customer.Email,
            EmailSubject = "Order Confirmation"
        };
    }
}

File: golang/02-never-ending-function/order_processor.go
package kata

import (
	"fmt"
	"math"
	"time"
)

type Item struct {
	Price    float64
	Quantity int
}
type Customer struct{ Email string }
type Address struct{ Zip string }
type Order struct {
	Items    []Item
	Customer Customer
	Address  Address
}
type OrderResult struct {
	Error         string
	Total         float64
	Carrier       string
	PaymentStatus string
	EmailTo       string
	OrderID       string
}

type OrderProcessor struct{}

func NewOrderProcessor() *OrderProcessor {
	return &OrderProcessor{}
}

func (op *OrderProcessor) process_order(order Order) OrderResult {
	if len(order.Items) == 0 {
		return OrderResult{Error: "No items"}
	}
	if order.Customer.Email == "" {
		return OrderResult{Error: "Invalid customer"}
	}
	if order.Address.Zip == "" {
		return OrderResult{Error: "Invalid address"}
	}
	inventory := true
	for _, item := range order.Items {
		if item.Quantity > 100 {
			inventory = false
		}
	}
	if !inventory {
		return OrderResult{Error: "Out of stock"}
	}
	subtotal := 0.0
	weight := 0
	for _, item := range order.Items {
		subtotal += item.Price * float64(item.Quantity)
		weight += item.Quantity
	}
	_ = weight
	shipping := 0.0
	if subtotal < 25 {
		shipping = 5.99
	} else if subtotal < 50 {
		shipping = 3.99
	}
	total := subtotal + subtotal*0.07 + shipping
	carrier := "USPS"
	if total > 100 {
		carrier = "UPS"
	}
	status := "approved"
	if total > 1000 {
		status = "pending_review"
	}
	return OrderResult{Total: math.Round(total*100) / 100, Carrier: carrier, PaymentStatus: status, EmailTo: order.Customer.Email, OrderID: fmt.Sprintf("ORD-%d", time.Now().UnixMilli())}
}
